datastore/longhorn.go

```diff
@@ -1167,17 +1167,210 @@ func (s *DataStore) ListBackingImages() (map[string]*longhorn.BackingImage, erro
 	return itemMap, nil
 }
 
-// GetOwnerReferencesForBackingImage returns OwnerReference for the given
-// backing image name and UID
-func GetOwnerReferencesForBackingImage(backingImage *longhorn.BackingImage) []metav1.OwnerReference {
+// CreateBackingImageManager creates a Longhorn BackingImageManager resource and verifies
+// creation
+func (s *DataStore) CreateBackingImageManager(backingImageManager *longhorn.BackingImageManager) (*longhorn.BackingImageManager, error) {
+	if err := util.AddFinalizer(longhornFinalizerKey, backingImageManager); err != nil {
+		return nil, err
+	}
+	if err := tagNodeLabel(backingImageManager.Spec.NodeID, backingImageManager); err != nil {
+		return nil, err
+	}
+	if err := tagDiskUUIDLabel(backingImageManager.Spec.DiskUUID, backingImageManager); err != nil {
+		return nil, err
+	}
+
+	ret, err := s.lhClient.LonghornV1beta1().BackingImageManagers(s.namespace).Create(backingImageManager)
+	if err != nil {
+		return nil, err
+	}
+	if SkipListerCheck {
+		return ret, nil
+	}
+
+	obj, err := verifyCreation(backingImageManager.Name, "backing image manager", func(name string) (runtime.Object, error) {
+		return s.getBackingImageManagerRO(name)
+	})
+	if err != nil {
+		return nil, err
+	}
+	ret, ok := obj.(*longhorn.BackingImageManager)
+	if !ok {
+		return nil, fmt.Errorf("BUG: datastore: verifyCreation returned wrong type for backing image manager")
+	}
+
+	return ret.DeepCopy(), nil
+}
+
+// UpdateBackingImageManager updates Longhorn BackingImageManager and verifies update
+func (s *DataStore) UpdateBackingImageManager(backingImageManager *longhorn.BackingImageManager) (*longhorn.BackingImageManager, error) {
+	if err := util.AddFinalizer(longhornFinalizerKey, backingImageManager); err != nil {
+		return nil, err
+	}
+	if err := tagNodeLabel(backingImageManager.Spec.NodeID, backingImageManager); err != nil {
+		return nil, err
+	}
+	if err := tagDiskUUIDLabel(backingImageManager.Spec.DiskUUID, backingImageManager); err != nil {
+		return nil, err
+	}
+
+	obj, err := s.lhClient.LonghornV1beta1().BackingImageManagers(s.namespace).Update(backingImageManager)
+	if err != nil {
+		return nil, err
+	}
+	verifyUpdate(backingImageManager.Name, obj, func(name string) (runtime.Object, error) {
```

Comment:
Right now there is no extra error handling for `verifyUpdate()` in all CR update functions. If it's required, we can create a separate ticket and modify them together. Same below.

Reply:
Sounds good

Reply:
I think this is a problem and we need to create a ticket to handle error when `verifyUpdate()` fails. For example, when the Longhorn successfully [update the backing image CR in etcd](https://github.com/longhorn/longhorn-manager/blob/08c89c2d5e36f5585df7939bbbdb94fe47562f45/datastore/longhorn.go#L1107) but fails to [verifyUpdate()](https://github.com/longhorn/longhorn-manager/blob/08c89c2d5e36f5585df7939bbbdb94fe47562f45/datastore/longhorn.go#L1111), it could mean the updated CR is not synced from etcd to controllers' cache yet. In the next reconciliation loop, we might [recreate the disk UUID ](https://github.com/longhorn/longhorn-manager/blob/08c89c2d5e36f5585df7939bbbdb94fe47562f45/controller/backing_image_controller.go#L225-L227) again

Reply:
The ticket is posted: https://github.com/longhorn/longhorn/issues/2456